### PR TITLE
[IMPROVE] Mark threads as read when calling `handleToggleRead`

### LINKED ISSUE #19077: Thread message notification does not disappear
### Description:

Once a thread message is appearing, I need to click on it in order to remove the blue notification picture. There is no other way to hide this notification.. In previous releases pressing ESC key removed all notifications from the current channel.
Now, at least under Rocketchat release 3.6.1 there is no other way to clean this notification.

Please note that I am using the following feature :
"Show thread messages in the main thread". Therefore, I am always able to see this "hidden" messages.

### Steps to reproduce:

1.    ask someone to reply to a message in a "thread"
2.   try to remove this annoying notification either by opening this channel or by pressing ESC in this channel

### Expected behavior:

as soon as I get to read this message, the notification should desapear.
Alright, hard to define. Let's make it simple: hitting the ESC key should clear all current channel notifications

### Actual behavior:

This blue notificatiion asside the channel name stay. It only disappear when I click on the actual "new" message

### Server Setup Information:

- Version of Rocket.Chat Server: 3.6.1
- Operating System: 
- Deployment Method: 
- Number of Running Instances: 1
- DB Replicaset Oplog: 
- NodeJS Version: 
- MongoDB Version:

### Client Setup Information

- Desktop App or Browser Version: Firefox 78.0.2
- Operating System: Fedora 33

### LINKED ISSUE #20657: Message counter isn't disappearing after reading the message
### Description:

I faced a annoying bug. A notification counter appeared when somebody chatted to me. After reading the message, the counter didn't disappear, and even right-clicking and marking as read, the counter still is there.
image

image

If that person speaks to me, the counter increases, but when I enter the conversation, the counter sets to "1" again.
If tried to reload server, reload page without cache, and opened desktop and other browsers, that counter is on everywhere.

### Steps to reproduce:

There is no way to reproduce it, is just this conversation what remains as unread

### Expected behavior:

The counter should disappear, or at least, when I mark the conversation as read, it should force counter to be reset.

### Actual behavior:

Counter stays at "1"

### Server Setup Information:

- Version of Rocket.Chat Server: 3.11.0
- Operating System: Debian 10
- Deployment Method: source
- Number of Running Instances:  1
- NodeJS Version: 12.18.2
- MongoDB Version: 4.0.19 

### Client Setup Information

- Desktop App or Browser Version: all versions
- Operating System: all OSes

### Relevant logs:

```
I20210209-14:01:11.896(1) server.js:204 Meteor ➔ method readMessages -> userId: Qjvc83g9tPWkS4fti, arguments: ["NuXEcqxXRkwwwMzxAQjvc83g9tPWkS4fti"] I20210209-14:01:14.192(1) server.js:204 Meteor ➔ method readMessages -> userId: Qjvc83g9tPWkS4fti, arguments: ["NuXEcqxXRkwwwMzxAQjvc83g9tPWkS4fti"] I20210209-14:01:15.708(1) server.js:204 Meteor ➔ method readMessages -> userId: Qjvc83g9tPWkS4fti, arguments: ["NuXEcqxXRkwwwMzxAQjvc83g9tPWkS4fti"] 
```

### LINKED ISSUE #22802: Mark read in chat menu doesn't affect unread threads
### Description:

**Mark read** option in chat menu doesn't mark chat as read if there is unread messages in threads

### Steps to reproduce:

1. Get unread message in thread in some chat
2. Try to mark that chat read
3. Chat still unread and with count in blue circle

### Expected behavior:

Chat become mark as read

### Actual behavior:

Here is a chat with unread message in thread
image
Try to mark room read
image
Even if user open chat and read all main chat messages there still count for unread thread
image

### Server Setup Information:

- Version of Rocket.Chat Server: 3.16.3
- Operating System: CentOS7
- Deployment Method: docker
- Number of Running Instances: 20
- DB Replicaset Oplog: Enabled
- NodeJS Version: 12.22.1
- MongoDB Version: 4.2.14

### Client Setup Information

- Desktop App or Browser Version: 3.2.3 and Chrome 92
- Operating System: Windows 10

### Additional context

I understand that thread unread messages it's a bit different thing then main chat messages.
But users need a button to mark chat as read despite of threads.
Maybe some option in users preferences like "Mark threads read too on mark chat as read" or something like that.

Now it's not a but, but it's not complete work of **Mark read** button, which must be improved.

### Relevant logs:

No

### PR DESCRIPTION
**Pull request by sidmohanty11**
Only if any unread threads are present an extra "Mark Threads Read" option is shown in the Room Menu.
Added a separate function that can filter out the unread messages in that specific room/channel(that is being clicked) and marks every thread as read. 

Please review this PR @MartinSchoeler @gabriellsh @debdutdeb and suggest for required changes.
Thank you.

## Proposed changes (including videos or screenshots)

mark-threads-read

## Issue(s)

Fix #22802
Fix #20657
Fix #19077

### CODE REVIEW

**Comment by ankar84** [comment]
Nice improvement!

**Comment by ankar84** [comment]
Waiting for that PR merged

**Comment by ankar84** [comment]
Any updates?

**Comment by matheusbsilva137** [comment]
> Any updates?

Hey @ankar84 , this is now being reviewed by our team! I expect to include this is the next release candidate 🙂

File: apps/meteor/client/sidebar/RoomMenu.js [hunk 1]
```diff
@@ -139,6 +141,24 @@ const RoomMenu = ({ rid, unread, threadUnread, alert, roomOpen, type, cl, name =
...
+		try {
+			if (threadUnread) {
+				const { threads } = await getThreadsList({
+					rid,
+					type: 'unread',
+				});
+				const promises = [];
+				for (const thread of threads) {
+					promises.push(readThreads(thread?._id));
+				}
```
Comment (matheusbsilva137) [inline apps/meteor/client/sidebar/RoomMenu.js:154]:
```suggestion
				const promises = threads.map(thread => readThreads(thread?._id));
```
Maybe we could use `map` here to improve legibility

**Review by matheusbsilva137 — CHANGES_REQUESTED** [review]

File: apps/meteor/client/sidebar/RoomMenu.js (hunk 1 above)
Comment (sidmohanty11) [reply apps/meteor/client/sidebar/RoomMenu.js]: yup, sure! done :+1:

**Comment by LucianoPierdona** [comment]
Hey @sidmohanty11, I've implemented a small refactor to mark threads read inside the existing mark read after discussions with the team, could you update this branch with the lastest changes so we can follow up on this? Thanks!

cc @matheusbsilva137

**Comment by sidmohanty11** [comment]
Hey @LucianoPierdona, sure. there are few errors thrown due to ts. 
```javascript
const { threads } = await getThreadsList({ rid, type: 'unread' });
```
this errors out because we need to pass in `offset` and `count`

and

```javascript
const readThreads = useMethod('readThreads');
```

errors out because `'"readThreads"' is not assignable to parameter of type 'keyof ServerMethods'`.

Do you have any suggestions?

**Comment by sidmohanty11** [comment]
@LucianoPierdona please have a look. I've made the changes accordingly. Fixed some other eslint warnings too. Please let me know if something needs to be changed.

Thanks!

**Review by carlosrodrigues94 — CHANGES_REQUESTED** [review]

**Comment by ankar84** [comment]
Any chances that fix lands on 5.1.0 release?

File: apps/meteor/client/sidebar/RoomMenu.tsx [hunk 2]
```diff
@@ -50,7 +51,9 @@ const RoomMenu = ({ rid, unread, threadUnread, alert, roomOpen, type, cl, name =
 
 	const subscription = useUserSubscription(rid, fields);
 	const canFavorite = useSetting('Favorite_Rooms');
-	const isFavorite = Boolean(subscription?.f);
+	const isFavorite = (subscription != null ? subscription.f : undefined) != null && subscription?.f;
```
Comment (gabriellsh) [inline apps/meteor/client/sidebar/RoomMenu.tsx:54]: why 😢

File: apps/meteor/client/sidebar/RoomMenu.tsx [hunk 3]
```diff
@@ -137,16 +141,35 @@ const RoomMenu = ({ rid, unread, threadUnread, alert, roomOpen, type, cl, name =
...
+	const handleThreadsToBeRead = useMutableCallback(async () => {
+		try {
+			if (threadUnread) {
+				const { threads } = await getThreadsList({
+					rid,
+					type: 'unread',
+					count: 0,
+					offset: 0,
+				});
+				const promises = threads.map((thread) => readThreads(thread?._id));
```
Comment (gabriellsh) [inline apps/meteor/client/sidebar/RoomMenu.tsx:153]: I'm wondering here if this should be front-end responsibility... Perhaps we should update the backend `readMessages` function to do this? Or even create a new endpoint for it if that's the case.

**Review by gabriellsh — CHANGES_REQUESTED** [review]

File: apps/meteor/client/sidebar/RoomMenu.tsx [hunk 4]
```diff
@@ -137,16 +141,35 @@ const RoomMenu = ({ rid, unread, threadUnread, alert, roomOpen, type, cl, name =
...
+				await Promise.all(promises);
+			}
+		} catch (error) {
+			dispatchToastMessage({ type: 'error', message: String(error) });
```
[the remark this answers is not in the document]
Comment (gabriellsh) [reply apps/meteor/client/sidebar/RoomMenu.tsx:157]: Yep. Also, `dispatchToastMessage` function's `message` argument receives an error object, and extracts the message automatically. Only thing that should be done here is do type assertion/narrowing.

**Comment by gabriellsh** [comment]
> I'm wondering here if this should be front-end responsibility

It really isn't. I'll be closing this PR since it has conflicts and the solution could case problems in the future. I appreciate your contribution, and please, feel free to open a new PR with the advices taken into account. Just closing because it doesn't make sense to reuse this branch.

Again, thanks a lot for your time!

**Pull request closed without merging**